ratis-shell/src/main/java/org/apache/ratis/shell/cli/sh/command/AbstractRatisCommand.java

```diff
@@ -0,0 +1,179 @@
+    String peersStr = "";
+    if (cl.hasOption(PEER_OPTION_NAME)) {
+      peersStr = cl.getOptionValue(PEER_OPTION_NAME);
+    }
+    String[] peersArray = peersStr.split(",");
+    for (int i = 0; i < peersArray.length; i++) {
+      String[] hostPortPair = peersArray[i].split(":");
+      InetSocketAddress addr =
+          new InetSocketAddress(hostPortPair[0], Integer.parseInt(hostPortPair[1]));
+      addresses.add(addr);
+    }
+
+    RaftGroupId raftGroupIdFromConfig = DEFAULT_RAFT_GROUP_ID;
+    if (cl.hasOption(GROUPID_OPTION_NAME)) {
+      raftGroupIdFromConfig = RaftGroupId.valueOf(
+          UUID.fromString(cl.getOptionValue(GROUPID_OPTION_NAME)));
+    }
+
+    List<RaftPeer> peers = addresses.stream()
+        .map(addr -> RaftPeer.newBuilder()
+            .setId(RaftUtils.getPeerId(addr))
+            .setAddress(addr)
+            .build()
+        ).collect(Collectors.toList());
+    raftGroup = RaftGroup.valueOf(raftGroupIdFromConfig, peers);
+    try (final RaftClient client = RaftUtils.createClient(raftGroup)) {
+      RaftGroupId remoteGroupId;
+      List<RaftGroupId> groupIds;
+      groupIds = RetryUtil.run(peers,
+          p -> {
+            try {
+              return client.getGroupManagementApi((p.getId())).list().getGroupIds();
+            } catch (IOException e) {
+              e.printStackTrace();
+              return null;
+            }
+          }
+      );
+
+      if (groupIds.size() == 1) {
+        remoteGroupId = groupIds.get(0);
+      } else {
+        final UUID raftGroupUuid = raftGroupIdFromConfig.getUuid();
+        Optional<RaftGroupId> raftGroupId =
+            groupIds.stream().filter(r -> raftGroupUuid.equals(r.getUuid()))
+                .findFirst();
+        if (!raftGroupId.isPresent()) {
+          printStream.println(
+              "there are more than one group, you should specific one."
+                  + groupIds);
+          return -1;
+        } else {
+          remoteGroupId = raftGroupId.get();
+        }
+      }
+      raftGroup = RetryUtil.run(peers,
+          p -> {
+            try {
+              return client.getGroupManagementApi((p.getId()))
+                  .info(remoteGroupId).getGroup();
+            } catch (IOException e) {
+              e.printStackTrace();
+              return null;
+            }
+          }
+      );
+    }
+    return 0;
+  }
+
+  @Override
+  public void validateArgs(CommandLine cl) throws IllegalArgumentException {
+    if (!cl.hasOption(PEER_OPTION_NAME)) {
+      throw new IllegalArgumentException(String.format(
+          "should provide [%s]", PEER_OPTION_NAME));
+    }
+  }
+
+  @Override
+  public Options getOptions() {
+    return new Options()
+            .addOption(PEER_OPTION_NAME, true, "Peer addresses seperated by comma")
+            .addOption(GROUPID_OPTION_NAME, true, "Raft group id");
+  }
+
+  protected PrintStream getPrintStream() {
+    return printStream;
+  }
```

Comment:
Replace getPrintStream() with printf and println below.
```
  protected void printf(String format, Object... args) {
    printStream.printf(format, args);
  }

  protected void println(Object message) {
    printStream.println(message);
  }
```

Reply:
I guess you mean we should't expose the printStream outside of `AbstractRatisCommand`? If so, it make sense.

Reply:
My intention is to make the code shorter.  :)